app/uk/gov/hmrc/gform/sharedmodel/SubmissionReferenceUtil.scala

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.gform.sharedmodel
+import java.security.MessageDigest
+import java.math.BigInteger
+import scala.math.pow
+
+import uk.gov.hmrc.gform.sharedmodel.form.EnvelopeId
+
+object SubmissionReferenceUtil {
+
+  def findDigits(source: Long, no: Int, digits: Array[Long]): Array[Long] =
+    if (no == 0) {
+      (source % 36) +: digits
+    } else if (no == 1) {
+      findDigits(source, no - 1, ((source / 36) % 36) +: digits)
+    } else {
+      findDigits(source, no - 1, ((source / pow(36L, no)) % 36).toLong +: digits)
+    }
+
+  def getSubmissionReference(envelopeId: EnvelopeId) = {
+    val digest = MessageDigest.getInstance("SHA-256").digest(envelopeId.value.getBytes()).take(8)
+    def toUnsigned(x: Byte): Long = if (x < 0) 127 - x else x
+    val x
+      : Long = toUnsigned(digest(0)) + 256 * (toUnsigned(digest(1)) + 256 * (toUnsigned(digest(2)) + 256 * (toUnsigned(
+      digest(3)) + 256 * (toUnsigned(digest(4)) + 256 * (toUnsigned(digest(5)) + 256 * (toUnsigned(digest(6)) + 256 * (toUnsigned(
```

Comment:
You will find this easier using << to shift left multiples of 8 and use &  0xff or and 0x7f to make out the unsigned part.  The combine that with say List.range(0, 6) and map over the digest byte array.  There might just be a little wrinkle over & being defined on ints but not on bytes.

Reply:
Actually, I think you can use BigInt to form an integer from the array of bytes, avoiding all of this.

E.g.:

import scala.math.BigInt
val byteArray = BigInt(1337).toByteArray
val int = BigInt(byteArray)